### PR TITLE
[MI-2261]:Created delete sub command for confluence config

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**

### CODE REVIEW

File: server/command.go [hunk 1]
```diff
@@ -155,14 +156,18 @@ func getAutoCompleteData() *model.AutocompleteData {
...
-		HelpText: "List all the added configs",
-		Item:     "list",
-	}}
-	config.AddStaticListArgument("", false, configOptions)
+
+	addConfig := model.NewAutocompleteData("add", "", "Add config for the confluence instance")
+
+	listConfig := model.NewAutocompleteData("list", "", "List all the added configs")
+
+	deleteConfig := model.NewAutocompleteData("delete", "[instance]", "Add config for the confluence instance")
```
Comment (ayusht2810) [inline server/command.go]:
```suggestion
	deleteConfig := model.NewAutocompleteData("delete", "[instance]", "Delete config for the confluence instance")
```

File: server/config_list.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	"encoding/json"
+	"net/http"
+	"strings"
+
+	"github.com/mattermost/mattermost-plugin-confluence/server/utils"
+	"github.com/mattermost/mattermost-server/v6/model"
+)
+
+func (p *Plugin) handleGetConfigList(w http.ResponseWriter, r *http.Request) {
+	userID := r.FormValue("user_id")
```
Comment (ayusht2810) [inline server/config_list.go]: constant for this

File: server/config_list.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	if !utils.IsSystemAdmin(userID) {
+		http.Error(w, "user is not system admin", http.StatusInternalServerError)
```
Comment (ayusht2810) [inline server/config_list.go]: should this be `StatusInternalServerError` or 401?

File: server/config_list.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+	out := []model.AutocompleteListItem{}
+	for _, key := range configKeys {
+		key = strings.Split(key, "_")[0]
+		key = strings.TrimSuffix(key, "/")
+		out = append(out, model.AutocompleteListItem{
+			Item: key,
+		})
+	}
+	b, _ := json.Marshal(out)
```
Comment (ayusht2810) [inline server/config_list.go]: give proper name

Comment (ayusht2810) [inline server/config_list.go]:
```suggestion
	}

	b, _ := json.Marshal(out)
```

File: server/instance_cloud.go [hunk 5]
```diff
@@ -31,7 +31,7 @@ func (p *Plugin) installCloudInstance(rawURL string) (string, *cloudInstance, er
 		InstanceCommon: newInstanceCommon(p, CloudInstanceType, types.ID(confluenceURL)),
 	}
 
-	err = p.InstallInstance(instance)
+	err = p.InstallInstance(instance, false)
 	if err != nil {
 		return "", nil, err
 	}
```
Comment (ayusht2810) [inline server/instance_cloud.go]:
```suggestion
	if err = p.InstallInstance(instance, false); err != nil {
		return "", nil, err
	}
```

File: server/instance_server.go [hunk 6]
```diff
@@ -28,7 +28,7 @@ func (p *Plugin) installServerInstance(rawURL string) (string, *serverInstance,
 		InstanceCommon: newInstanceCommon(p, ServerInstanceType, types.ID(confluenceURL)),
 	}
 
-	err = p.InstallInstance(instance)
+	err = p.InstallInstance(instance, false)
 	if err != nil {
 		return "", nil, err
 	}
```
Comment (ayusht2810) [inline server/instance_server.go]: same

File: server/kv.go [hunk 7]
```diff
@@ -265,6 +266,22 @@ func (store *store) LoadSavedConfigs(configKeys []string) ([]*serializer.Conflue
...
+		if returnErr == nil {
+			return
+		}
+		returnErr = errors.WithMessage(returnErr,
+			fmt.Sprintf("failed to delete config for instance:%s", instanceID))
+	}()
+	appErr := store.plugin.API.KVDelete(keyWithInstanceIDForConfig(instanceID))
+	if appErr != nil {
+		return appErr
+	}
```
Comment (ayusht2810) [inline server/kv.go]: same

File: server/kv_mock_test.go [hunk 8]
```diff
@@ -130,6 +130,9 @@ func (store mockInstanceStore) LoadInstanceConfig(string) (*serializer.Confluenc
 func (store mockInstanceStore) LoadSavedConfigs([]string) ([]*serializer.ConfluenceConfig, error) {
 	return []*serializer.ConfluenceConfig{}, nil
 }
+func (store mockInstanceStore) DeleteInstanceConfig(string) error {
```
Comment (ayusht2810) [inline server/kv_mock_test.go]:
```suggestion
}

func (store mockInstanceStore) DeleteInstanceConfig(string) error {
```

File: server/plugin.go [hunk 9]
```diff
@@ -260,12 +257,6 @@ func (p *Plugin) OnInstall(c *plugin.Context, event model.OnInstallEvent) error
 
 func (c *externalConfig) processConfiguration() error {
```
Comment (ayusht2810) [inline server/plugin.go]: do we need to return `error` from here?

File: server/user.go [hunk 10]
```diff
@@ -248,7 +248,7 @@ func (p *Plugin) CompleteOAuth2(mattermostUserID, code, state string, instance I
 		ci := instance.(*cloudInstance)
 		ci.CloudID = cloudID
 		// Update the instance stored in the store
-		err = p.InstallInstance(ci)
+		err = p.InstallInstance(ci, true)
 		if err != nil {
 			return nil, nil, err
```
Comment (ayusht2810) [inline server/user.go]: same

File: server/command.go [hunk 11]
```diff
@@ -354,6 +372,22 @@ func listConfig(p *Plugin, context *model.CommandArgs, args ...string) *model.Co
...
 }
 
+func deleteConfig(p *Plugin, context *model.CommandArgs, args ...string) *model.CommandResponse {
+	if !utils.IsSystemAdmin(context.UserId) {
+		p.postCommandResponse(context, installOnlySystemAdmin)
+		return &model.CommandResponse{}
+	}
+
+	instance := strings.Join(args, " ")
+
```
Comment (raghavaggarwal2308) [inline server/command.go]:
```suggestion
```

File: server/kv.go [hunk 12]
```diff
@@ -265,6 +266,21 @@ func (store *store) LoadSavedConfigs(configKeys []string) ([]*serializer.Conflue
 	return configs, nil
 }
 
+func (store *store) DeleteInstanceConfig(instanceID string) (returnErr error) {
+	defer func() {
+		if returnErr == nil {
+			return
+		}
+		returnErr = errors.WithMessage(returnErr,
+			fmt.Sprintf("failed to delete config for instance:%s", instanceID))
```
Comment (raghavaggarwal2308) [inline server/kv.go]:
```suggestion
			fmt.Sprintf("failed to delete config for instance: %s", instanceID))
```

**Review by raghavaggarwal2308 — APPROVED** [review]

File: server/plugin.go [hunk 13]
```diff
@@ -255,9 +252,8 @@ func (p *Plugin) OnInstall(c *plugin.Context, event model.OnInstallEvent) error
 	return nil
 }
 
-func (c *externalConfig) processConfiguration() error {
+func (c *externalConfig) processConfiguration() {
```
Comment (ayusht2810) [inline server/plugin.go:249]: shouldn't we remove this function and use the single line below directly?

**Review by ayusht2810 — APPROVED** [review]

File: server/config_list.go [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+	"github.com/mattermost/mattermost-plugin-confluence/server/utils"
+	"github.com/mattermost/mattermost-server/v6/model"
+)
+
+const paramUserID = "user_id"
+
+func (p *Plugin) handleGetConfigList(w http.ResponseWriter, r *http.Request) {
+	userID := r.FormValue(paramUserID)
+	if !utils.IsSystemAdmin(userID) {
+		http.Error(w, "user is not system admin", http.StatusUnauthorized)
```
Comment (avas27JTG) [inline server/config_list.go]:
```suggestion
		http.Error(w, "user is not a system admin", http.StatusUnauthorized)
```

**Review by avas27JTG — APPROVED** [review]

**Review by avas27JTG — APPROVED** [review]

**Review by avas27JTG — APPROVED** [review]

**Pull request merged**